Hardening and robustness fixes

mgerstner opened the pull request:
These hardenings and improvements are the result of a code review for inclusion of efs-utils in SUSE Linux Enterprise products.

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -385,13 +385,25 @@ def start_watchdog(init_system):
         sys.stderr.write('%s\n' % error_message)
         logging.warning(error_message)
 
```
two blank lines between methods please

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -385,13 +385,25 @@ def start_watchdog(init_system):
+        pass
+
+    os.makedirs(state_file_dir, mode)
 
```
two blank lines between methods please

patelia commented on src/watchdog/__init__.py:
```diff
@@ -95,6 +95,9 @@ def get_file_safe_mountpoint(mount):
         mountpoint = mountpoint[1:]
 
     opts = parse_options(mount.options)
+    if not 'port' in opts:
```
`if 'port' not in opts:` is the preferred python way

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -364,19 +364,19 @@ def check_network_status(fs_id, init_system):
 def start_watchdog(init_system):
     if init_system == 'init':
-        proc = subprocess.Popen(['/sbin/status', WATCHDOG_SERVICE], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        proc = subprocess.Popen(['/sbin/status', WATCHDOG_SERVICE], stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
```
please break into two lines, we wrap lines at 130 characters

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -404,7 +404,7 @@ def bootstrap_tls(config, init_system, dns_name, fs_id, mountpoint, options, sta
     # launch the tunnel in a process group so if it has any child processes, they can be killed easily by the mount watchdog
     logging.info('Starting TLS tunnel: "%s"', ' '.join(tunnel_args))
-    tunnel_proc = subprocess.Popen(tunnel_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, preexec_fn=os.setsid)
+    tunnel_proc = subprocess.Popen(tunnel_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, preexec_fn=os.setsid, close_fds=True)
```
please break into two lines, we line wrap at 130 characters

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -385,13 +385,25 @@ def start_watchdog(init_system):
+def create_state_file_dir(config, state_file_dir):
+    mode = 0o750
+    try:
+        mode_str = config.get(CONFIG_SECTION, "state_file_dir_mode")
```
please use single quotes for the string

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -385,13 +385,25 @@ def start_watchdog(init_system):
+        try:
+            mode = int(mode_str, 8)
+        except ValueError:
+            logging.warn('bad state_file_dir_mode "%s" in config file', mode_str)
```
Capitalize the "b" in "Bad", and log the config file path as well (defined at the top as `CONFIG_FILE`)

mgerstner commented:
I've adjusted and rebased according to your requests.

mgerstner commented:
> Please also ensure that the changes pass unit tests and flake8. Currently the `test_bootstrap_tls_state_file_nonexistent_dir` test is failing:
> 
> ```
>     def create_state_file_dir(config, state_file_dir):
>         mode = 0o750
>         try:
>             mode_str = config.get(CONFIG_SECTION, "state_file_dir_mode")
>             try:
> >               mode = int(mode_str, 8)
> E               TypeError: int() can't convert non-string with explicit base
> 
> src/mount_efs/__init__.py:396: TypeError
> ```

This is a strange one, the error is actually caused by the Mock setup in the test itself, causing the `config.get()` not to return a `str` but some `MagicMock` type. I'm not very familiar with pytest so I've implemented a work around for this at the moment. If you know of some better way please tell me.

patelia commented on src/mount_efs/__init__.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env -S python -Es
```
From what i can tell, the version of `/usr/bin/env` that ships with the distros we support for `amazon-efs-utils` does not actually support the `-S` option. This causes the mount helper and watchdog to not launch properly. Simply removing the `-S` seems to cause other issues, but that may be an artifact of `python -Es` not being parsed correctly without the `-S` option. 

Is there an alternative way you could add `python -Es` without `/usr/bin/env -S`? Or a way to properly update `env`? We currently support `amazon-efs-utils` on the following distros:

* Amazon Linux
* Amazon Linux 2
* CentOS 7
* RHEL 7
* Debian 9
* Ubuntu 16.04

mgerstner replied:
Oh indeed the `-S` argument of /usr/bin/env is a rather recent addition to coreutils version 8.30 (released 2018-07-01).

This shebang business is a rather dark corner on Linux. The kernel does only minimal command line processing for the she-bang. It separates the interpreter from the rest of the command line and passes the command line as a single parameter to the interpreter, ignoring any whitespace. The `-S` parameter to `/usr/bin/env` triggers userspace logic to split up the single argument received into separate ones.

So I see various approaches to deal with this:

- Drop `/usr/bin/env` and simply use `#!/usr/bin/python -Es` as shebang. This works if all of the targeted distributions install python in `/usr/bin/python` which I guess is pretty common.
- Determine the correct location of `python` during build time and fill in the result as shebang in the Python scripts.
- Use a wrapper shell script with shebang `#!/bin/sh` that in turn runs `exec /usr/bin/env python -Es /path/to/python-script`. This would require changes in the installation logic of efs-utils.
- There's also the possiblity to use a crude hack:
```
#!/bin/sh

""":";exec /usr/bin/env python -Es $0;":"""

import os
import sys
```

The hack abuses the python docstring, to be interpreted by the regular shell, restarting itself as the python interpreter with the desired lookup semantics and command line arguments.

patelia commented:
> > Please also ensure that the changes pass unit tests and flake8. Currently the `test_bootstrap_tls_state_file_nonexistent_dir` test is failing:
> > ```
> >     def create_state_file_dir(config, state_file_dir):
> >         mode = 0o750
> >         try:
> >             mode_str = config.get(CONFIG_SECTION, "state_file_dir_mode")
> >             try:
> > >               mode = int(mode_str, 8)
> > E               TypeError: int() can't convert non-string with explicit base
> > 
> > src/mount_efs/__init__.py:396: TypeError
> > ```
> 
> This is a strange one, the error is actually caused by the Mock setup in the test itself, causing the `config.get()` not to return a `str` but some `MagicMock` type. I'm not very familiar with pytest so I've implemented a work around for this at the moment. If you know of some better way please tell me.

I think you want to mock the specific call to get the `state_file_dir_mode` from the config, which is [done with `side_effect`](https://docs.python.org/3/library/unittest.mock.html#calling). So the test would look this:

```
def test_bootstrap_tls_state_file_nonexistent_dir(mocker, tmpdir):
    _mock_popen(mocker)
    mocker.patch('os.kill')
    state_file_dir = str(tmpdir.join(tempfile.mktemp()))

    def config_get_side_effect(section, field):
        if section == mount_efs.CONFIG_SECTION and field == 'state_file_dir_mode':
            return '0755'
        else:
            raise ValueError('Unexpected arguments')

    MOCK_CONFIG.get.side_effect = config_get_side_effect

    assert not os.path.exists(state_file_dir)

    with mount_efs.bootstrap_tls(MOCK_CONFIG, INIT_SYSTEM, DNS_NAME, FS_ID, MOUNT_POINT, {}, state_file_dir):
        pass

    assert os.path.exists(state_file_dir)
```

mgerstner commented:
> I think you want to mock the specific call to get the `state_file_dir_mode` from the config, which is [done with `side_effect`](https://docs.python.org/3/library/unittest.mock.html#calling). So the test would look this:

Thank you for the input. I've adjusted and rebased by PR accordingly.

mgerstner commented:
Since the issue with the she-bang line is more complicated I've dropped this commit from the PR such that the rest of the changes can make it through.

patelia commented:
With the removal of the shebang commit, the changes look good to me. Another PR was merged in the meantime, so I do have to ask you to update the version to 1.7 now. Please also rebase the `state_file_dir` and `pytest` commits into one, as they are related. 

I'll have another core contributor take a look at your changes.

mgerstner commented:
Sorry for the noise I had some troubles merging in the changes from your master branch and rebasing my commits. Things should be as requested by now.
